### ISSUE TITLE
REST API Guide lines should have authentication section

### ISSUE THREAD

**Issue by kyyberi** [APIOps/APIOps-cycles#13]
Although APIs will use different kind of authentication case by case, we should define the guide for that part as well. Currently authentication is not touched at all.

**Comment by mniinio** [comment]
Reason why authentication was not touched at all was that it needs a little bit of thinking in this "best practices" context. We should agree on the principles first.

All APIs should use OpenID Connect with JWT based claims as the principle because this is the safest, easiest and supported by most API management, tooling and framework providers

There are situations where OpenID Connect approach causes problems 
- not all analytics applications, legacy systems and marketing systems support OpenID Connect
- you need to have an identity provider and authentication service (either external or set up with tooling from the api management). If you are doing mobile apps, applications requiring development, if you have Azure AD or other SSO provider, chances are you need it anyway.
- OpenID Connect might feel like overkill to some scenarios, like having a completely public API with no other data in it than what is already publically in the internet. But even in that scenario, you do need authentication which is not just a dummy API key as a query parameter which gets logged and sniffed etc.

Using OpenID Connect and JWT with claims together with API management product and gateway to secure the microservice based APIs is obviously the thing to do. But it's also the stuff that requires plenty of explaining and consulting and has been explained elsewhere in the internet quite clearly. Also the API Audit referes to these documentations as well as OWASP REST cheat sheet and assumes that you have figured out the privacy and security needs AND done your API Consumer interviews as instructed in the APIOps Cycles. And that way you have come to the conclusion which is the best method for your API

**Comment by kyyberi** [comment]
Yes, I agree that authentication is tricky. However that is not valid reason to exclude one of the most important topic regarding APIs. We could introduce the most common approaches and best practices with those. Of course we can add that we are not liable if you follow the patterns and get hacked and that authentication is always solved case by case. 

Of course web is full of useful data. The whole APIOps Cycles could have been collection of links more most parts but still we decided to collect (and in some parts rewrite best practices found elsewhere). 

I would still consider having some kind of authentication section.

**Comment by mniinio** [comment]
I definitely didn't say we didn't need one, just that the approach needs to be thought first :D but I'll pick this one as this is one of the topics I deal with every day
